File: java/Trees/BinaryTree.java
package Trees;

import definitions.BTree.BTreeNode;

import java.util.*;

public class BinaryTree<T> {

    protected List<T> inOrder(BTreeNode<T> node) {
        List<T> ans = new ArrayList<>();
        if (node == null) return ans;
        ans.addAll(inOrder(node.left));
        ans.add(node.val);
        ans.addAll(inOrder(node.right));
        return ans;
    }

    protected List<T> preOrder(BTreeNode<T> node) {
        List<T> ans = new ArrayList<>();
        if (node == null) return ans;
        ans.add(node.val);
        ans.addAll(preOrder(node.left));
        ans.addAll(preOrder(node.right));
        return ans;
    }

    protected List<T> postOrder(BTreeNode<T> node) {
        List<T> ans = new ArrayList<>();
        if (node == null) return ans;
        ans.addAll(postOrder(node.left));
        ans.addAll(postOrder(node.right));
        ans.add(node.val);
        return ans;
    }

    protected List<List<T>> levelOrder(BTreeNode<T> node) {
        Map<Integer, List<T>> map = new LinkedHashMap<>();
        levelOrderHelper(node, map, 0);
        return map.values().stream().toList();

    }

    private void levelOrderHelper(BTreeNode<T> node, Map<Integer, List<T>> map, int level) {
        if (node == null) return;
        if (!map.containsKey(level)) map.put(level, new ArrayList<>());
        map.get(level).add(node.val);
        levelOrderHelper(node.left, map, level + 1);
        levelOrderHelper(node.right, map, level + 1);
    }


    protected List<T> leftView(BTreeNode<T> node) {
        var map = new LinkedHashMap<Integer, T>();
        leftViewHelper(node, map, 0);
        return map.values().stream().toList();
    }

    private void leftViewHelper(BTreeNode<T> node, Map<Integer, T> map, int level) {
        if (node == null) return;
        map.putIfAbsent(level, node.val);
        leftViewHelper(node.left, map, level + 1);
        leftViewHelper(node.right, map, level + 1);
    }

    protected List<T> rightView(BTreeNode<T> node) {
        var map = new LinkedHashMap<Integer, T>();
        rightViewHelper(node, map, 0);
        return map.values().stream().toList();
    }

    private void rightViewHelper(BTreeNode<T> node, Map<Integer, T> map, int level) {
        if (node == null) return;
        map.put(level, node.val);
        rightViewHelper(node.left, map, level + 1);
        rightViewHelper(node.right, map, level + 1);
    }

    protected int height(BTreeNode<T> node) {
        if (node == null) return 0;
        return Math.max(height(node.left), height(node.right)) + 1;
    }

    protected List<T> topView(BTreeNode<T> node) {
        var map = new TreeMap<Integer, T>();
        topViewHelper(node, map, 0);
        return map.values().stream().toList();
    }

    private void topViewHelper(BTreeNode<T> node, TreeMap<Integer, T> map, int scale) {
        if (node == null) return;
        map.putIfAbsent(scale, node.val);
        topViewHelper(node.left, map, scale - 1);
        topViewHelper(node.right, map, scale + 1);
    }

    protected List<T> bottomView(BTreeNode<T> node) {
        var map = new TreeMap<Integer, T>();
        bottomViewHelper(node, map, 0);
        return map.values().stream().toList();
    }

    private void bottomViewHelper(BTreeNode<T> node, TreeMap<Integer, T> map, int scale) {
        if (node == null) return;
        map.put(scale, node.val);
        bottomViewHelper(node.left, map, scale - 1);
        bottomViewHelper(node.right, map, scale + 1);
    }

    protected List<T> levelOrderArray(BTreeNode<T> node) {
        List<T> ans = new ArrayList<>();
        if (node == null) return ans;
        Queue<BTreeNode<T>> q = new LinkedList<>();
        q.add(node);
        while (!q.isEmpty()) {
            var elementNode = q.remove();
            ans.add(elementNode.val);
            if (elementNode.left != null) q.add(elementNode.left);
            if (elementNode.right != null) q.add(elementNode.right);
        }
        return ans;
    }

    protected List<List<T>> verticalOrder(BTreeNode<T> node) {
        var map = new TreeMap<Integer, List<T>>();
        verticalOrderHelper(node, map, 0);
        return map.values().stream().toList();
    }


    private void verticalOrderHelper(BTreeNode<T> node, Map<Integer, List<T>> map, int scale) {
        if (node == null) return;
        if (!map.containsKey(scale)) map.put(scale, new ArrayList<>());
        map.get(scale).add(node.val);
        verticalOrderHelper(node.left, map, scale - 1);
        verticalOrderHelper(node.right, map, scale + 1);

    }


    public static void main(String[] args) {
        var tree = new BinaryTree<Integer>();
        var root = new BTreeNode<Integer>(1);
        var a = new BTreeNode<Integer>(2);
        var b = new BTreeNode<Integer>(3);
        var c = new BTreeNode<Integer>(4);
        var d = new BTreeNode<Integer>(5);
        var e = new BTreeNode<Integer>(6);

        root.left = a;
        root.right = b;
        a.left = c;
        a.right = d;
        b.right = e;

        System.out.println(tree.inOrder(root));
        System.out.println(tree.preOrder(root));
        System.out.println(tree.postOrder(root));
        System.out.println(tree.levelOrder(root));
        System.out.println("Left view ->" + tree.leftView(root));
        System.out.println("Right view ->" + tree.rightView(root));
        System.out.println("Top view ->" + tree.topView(root));
        System.out.println("Bottom view ->" + tree.bottomView(root));
        System.out.println("Level order array ->" + tree.levelOrderArray(root));
        System.out.println("Vertical Order ->" + tree.verticalOrder(root));
    }
}

File: typescript/src/trees/binary-tree.ts
import Queue from "../queues/queue.ts";
import { AvlTreeNode } from "./avl-tree.ts";

export class BinaryTreeNode<T> {
  val: T;
  left: BinaryTreeNode<T> | null;
  right: BinaryTreeNode<T> | null;

  constructor(val: T) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

type BinaryTreeNodes<T> = BinaryTreeNode<T> | AvlTreeNode<T> | null;

export class BinaryTree<T> {
  public height(node: BinaryTreeNodes<T>): number {
    if (node == null) return 0;
    return Math.max(this.height(node.left), this.height(node.right)) + 1;
  }

  public inOrder(node: BinaryTreeNodes<T>) {
    const ans: T[] = [];
    if (!node) return ans;
    ans.push(...this.inOrder(node!.left));
    ans.push(node.val);
    ans.push(...this.inOrder(node!.right));
    return ans;
  }

  public preOrder(node: BinaryTreeNodes<T>) {
    const ans: T[] = [];
    if (!node) return ans;
    ans.push(node.val);
    ans.push(...this.preOrder(node!.left));
    ans.push(...this.preOrder(node!.right));
    return ans;
  }

  public postOrder(node: BinaryTreeNodes<T>) {
    const ans: T[] = [];
    if (!node) return ans;
    ans.push(...this.postOrder(node!.left));
    ans.push(...this.postOrder(node!.right));
    ans.push(node.val);
    return ans;
  }

  public levelOrder(node: BinaryTreeNodes<T>) {
    const map = new Map<number, T[]>();
    this.levelOrderHelper(node, map, 0);
    return [...map.values()];
  }

  private levelOrderHelper(
    node: BinaryTreeNodes<T>,
    levelMap: Map<number, T[]>,
    level: number,
  ) {
    if (node == null) return;
    if (!levelMap.has(level)) levelMap.set(level, []);
    levelMap.get(level)!.push(node.val);
    this.levelOrderHelper(node.left, levelMap, level + 1);
    this.levelOrderHelper(node.right, levelMap, level + 1);
  }

  public levelOrderArray(node: BinaryTreeNodes<T>) {
    const ans: T[] = [];
    if (node == null) return ans;
    const q = new Queue<BinaryTreeNode<T>>();
    q.add(node);

    while (q.size != 0) {
      const removed = q.remove()!;
      ans.push(removed.val);
      if (removed.left != null) q.add(node.left!);
      if (removed.right != null) q.add(node.right!);
    }

    return ans;
  }

  public leftView(
    node: BinaryTreeNodes<T>,
  ) {
    const map = new Map<number, T>();
    this.leftViewHelper(node, map, 0);
    return [...map.values()];
  }

  private leftViewHelper(
    node: BinaryTreeNodes<T>,
    map: Map<number, T>,
    level: number,
  ) {
    if (node == null) return;
    if (!map.has(level)) map.set(level, node.val);
    this.leftViewHelper(node.left, map, level + 1);
    this.leftViewHelper(node.right, map, level + 1);
  }

  public rightView(node: BinaryTreeNodes<T>) {
    const map = new Map<number, T>();
    this.rightViewHelper(node, map, 0);
    return [...map.values()];
  }

  private rightViewHelper(
    node: BinaryTreeNodes<T>,
    map: Map<number, T>,
    level: number,
  ) {
    if (node == null) return;
    map.set(level, node.val);
    this.leftViewHelper(node.left, map, level + 1);
    this.leftViewHelper(node.right, map, level + 1);
  }

  public topView(node: BinaryTreeNodes<T>) {
    const map = new Map<number, T>();
    this.topViewHelper(node, 0, map);
    return map.entries().toArray().sort((a, b) => a[0] - b[0]).map((it) =>
      it[1]
    );
  }

  private topViewHelper(
    node: BinaryTreeNodes<T>,
    scale: number,
    map: Map<number, T>,
  ) {
    if (node == null) return;
    if (!map.has(scale)) map.set(scale, node.val);
    this.topViewHelper(node.left, scale - 1, map);
    this.topViewHelper(node.right, scale + 1, map);
  }

  public bottomView(node: BinaryTreeNodes<T>) {
    const map = new Map<number, T>();
    this.bottomViewHelper(node, 0, map);
    return map.entries().toArray().sort((a, b) => a[0] - b[0]).map((it) =>
      it[1]
    );
  }

  private bottomViewHelper(
    node: BinaryTreeNodes<T>,
    scale: number,
    map: Map<number, T>,
  ) {
    if (node == null) return;
    if (!map.has(scale)) map.set(scale, node.val);
    this.topViewHelper(node.left, scale - 1, map);
    this.topViewHelper(node.right, scale + 1, map);
  }

  public verticalOrder(node: BinaryTreeNodes<T>) {
    const map = new Map<number, T[]>();
    this.verticalOrderHelper(node, 0, map);
    return map.entries().toArray().sort((a, b) => a[0] - b[0]).map((it) =>
      it[1]
    );
  }

  private verticalOrderHelper(
    node: BinaryTreeNodes<T>,
    scale: number,
    map: Map<number, T[]>,
  ) {
    if (node == null) return;
    if (!map.has(scale)) map.set(scale, []);
    map.get(scale)!.push(node.val);
    this.verticalOrderHelper(node.left, scale - 1, map);
    this.verticalOrderHelper(node.right, scale + 1, map);
  }
}
